### PR TITLE
Add `affine_transform` op to all backends

### PR DESCRIPTION
**Pull request by james77777778**
Related to #460 

This PR introduces the `affine` operation in 3 backends. This op can be useful when implementing KPLs using pure keras-core, such as RandomRotation, RandomTranslation, etc.

It is worth noting that the implementations of `F.affine_grid` and `F.grid_sample` in the torch backend differ significantly from TF and JAX. As a result, the correctness tests are expected to fail.

Hope this helps!

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR. It this really necessary? Couldn't you do the same with `einsum`?

**Comment by james77777778** [comment]
> Thanks for the PR. It this really necessary? Couldn't you do the same with `einsum`?

Do you mean that `affine` is not part of the plan, or should I implement the torch's `affine` operation using `einsum`?

For 1st:
It is not obvious for me to directly implement affine using `einsum`.
(actually, I have not seen the implementation of affine in these backends utilizes `einsum`?)

If the future plan involves implementing KPLs using pure keras-core, the `affine` op becomes neccesary for many layers.

For 2nd:
I can try to implement torch's `map_coordinate` by `einsum`.

**Comment by fchollet** [comment]
What's the closest numpy/scipy op? Does this match https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.affine_transform.html ?

**Comment by james77777778** [comment]
> What's the closest numpy/scipy op? Does this match https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.affine_transform.html ?

The challenge is that the absence of `map_coordinates` op in TF and Torch. And yes, I want to add `affine_transform` op to all backends!

In these 3 backends, JAX is the only one that can directly utilize numpy/scipy-like ops because it has `map_coordinates` op.

For TF, the closest op would be `tf.raw_ops.ImageProjectiveTransformV3`. This PR mainly follow the signature of `tf.raw_ops.ImageProjectiveTransformV3` and the transform definition of `[a0, a1, a2, b0, b1, b2, c0, c1]`

For Torch, the closest op is a combination of `tnn.affine_grid` and `tnn.grid_sample`.

**Comment by james77777778** [comment]
Usage:

```python
import matplotlib.cbook as cbook
import matplotlib.pyplot as plt
import numpy as np

from keras_core.backend.jax.image import affine as jax_affine
from keras_core.backend.tensorflow.image import affine as tf_affine
from keras_core.backend.torch.image import affine as torch_affine

with cbook.get_sample_data("grace_hopper.jpg") as image_file:
    image = plt.imread(image_file)
transform = np.array([1.2, 0.3, -100, 0.5, 1.2, -150, 0, 0])

jax_affined = jax_affine(image, transform)
tf_affined = tf_affine(image, transform)
torch_affined = torch_affine(image, transform)

fig, ax_dict = plt.subplot_mosaic([["A", "B", "C", "D"]], figsize=(12, 4))
ax_dict["A"].set_title("Original")
ax_dict["A"].imshow(image)
ax_dict["B"].set_title("JAX Affined")
ax_dict["B"].imshow(jax_affined)
ax_dict["C"].set_title("TF Affined")
ax_dict["C"].imshow(tf_affined)
ax_dict["D"].set_title("Torch Affined")
ax_dict["D"].imshow(torch_affined)
fig.tight_layout(h_pad=0.1, w_pad=0.1)
plt.savefig("affine.png")

```

The visualization:

affine

**Comment by fchollet** [comment]
> The challenge is that the absence of map_coordinates op in TF and Torch

@mihirparadkar is working on adding this exact op to Keras -- let's make sure he sees this.

My understanding was that torch's `grid_sample` was equivalent -- is it not?

Also, if we have `map_coordinates`, would we still need `affine`?

**Comment by james77777778** [comment]
> My understanding was that torch's `grid_sample` was equivalent -- is it not?

I believe `grid_sample` shares similar functionality but it is not precisely identical in terms of numerical behavior. 
(I don't know where the differences come from, but the correctness tests failed.)

Besides:
- `grid_sample` requires normalized coordinates [-1, 1] that is different from `map_coordinates`.
- `grid_sample` lacks the functionality to fill user-defined value. This PR try to add `fill_value` by modifying https://github.com/pytorch/vision/blob/main/torchvision/transforms/v2/functional/_geometry.py#L478

> Also, if we have `map_coordinates`, would we still need `affine`?

As a user, I think it would be more convenient to have `affine` or `affine_transform` in keras-core. It is not straightforward to apply affine transform using `map_coordinates`.

As a developer, `affine` is very useful to implement RandomRotation, RandomAffine, etc.

**Comment by fchollet** [comment]
> As a user, I think it would be more convenient to have affine or affine_transform in keras-core

Inversely, if we have `affine_transform`, do we still need `map_coordinates`?

**Comment by james77777778** [comment]
> Inversely, if we have `affine_transform`, do we still need `map_coordinates`?

For common 2D image processing, I believe that `affine_transform` is sufficient.
However, when it comes to the multidimensional image processing, `map_coordinates` might be the only option.

A use case for medical image processing:
https://bic-berkeley.github.io/psych-214-fall-2016/map_coordinates.html

**Comment by fchollet** [comment]
Understood. Let's add both `affine_transform` (same API as scipy, though we don't need to implement all interpolation modes) and `map_coordinates`. @mihirparadkar is working on `map_coordinates`.

File: keras_core/backend/jax/image.py [hunk 1]
```diff
@@ -40,3 +45,112 @@ def resize(
...
+    transform,
+    method="bilinear",
+    fill_mode="constant",
+    fill_value=0,
+    data_format="channels_last",
+):
+    if method not in AFFINE_METHODS.keys():
+        raise ValueError(
+            "Invalid value for argument `method`. Expected of one "
+            f"{AFFINE_METHODS.keys()}. Received: method={method}"
```
Comment (fchollet) [inline keras_core/backend/jax/image.py:74]: This will print "dict_keys()". Let's cast to `set()` first.

File: keras_core/ops/image.py [hunk 2]
```diff
@@ -102,3 +102,70 @@ def resize(
...
+@keras_core_export("keras_core.ops.image.affine")
+def affine(
+    image,
+    transform,
+    method="bilinear",
+    fill_mode="constant",
+    fill_value=0,
+    data_format="channels_last",
+):
+    # TODO: add docstring
```
Comment (fchollet) [inline keras_core/ops/image.py:156]: Please add the docstring

File: keras_core/backend/torch/image.py [hunk 3]
```diff
@@ -69,3 +72,149 @@ def resize(
...
+            # img * mask + (1.0 - mask) * fill =
+            # img * mask - fill * mask + fill =
+            # mask * (img - fill) + fill
+            float_img = float_img.sub_(fill_img).mul_(mask).add_(fill_img)
+
+    img = float_img.round_().to(img.dtype) if not fp else float_img
+    return img
+
+
+def affine(
```
Comment (fchollet) [inline keras_core/backend/torch/image.py:149]: Are you able to add it for the NumPy backend as well? Fine to depend on scipy.

**Review by fchollet — COMMENTED** [review]
Great work 👍

File: keras_core/backend/jax/image.py [hunk 4]
```diff
@@ -40,3 +45,112 @@ def resize(
...
+AFFINE_FILL_MODES = (
+    "constant",
+    "nearest",
+    "wrap",
+    "mirror",
+    "reflect",
+)
+
+
+def affine(
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_core/backend/jax/image.py]: Done

File: keras_core/backend/jax/image.py (hunk 1 above)
Comment (james77777778) [reply keras_core/backend/jax/image.py]: Done for all backends.

File: keras_core/ops/image.py (hunk 2 above)
Comment (james77777778) [reply keras_core/ops/image.py]: Added. I mainly borrowed the docstring from
- https://www.tensorflow.org/addons/api_docs/python/tfa/image/transform
- https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/image/apply_affine_transform

File: keras_core/backend/torch/image.py (hunk 3 above)
Comment (james77777778) [reply keras_core/backend/torch/image.py]: Added.

**Comment by james77777778** [comment]
Hi @fchollet 
I have added `affine_transform` to NumPy and it passed the tests as expected.

Usage:
```python
import matplotlib.cbook as cbook
import matplotlib.pyplot as plt
import numpy as np

from keras_core.backend.jax.image import (
    affine_transform as jax_affine_transform,
)
from keras_core.backend.numpy.image import (
    affine_transform as np_affine_transform,
)
from keras_core.backend.tensorflow.image import (
    affine_transform as tf_affine_transform,
)
from keras_core.backend.torch.image import (
    affine_transform as torch_affine_transform,
)

with cbook.get_sample_data("grace_hopper.jpg") as image_file:
    image = plt.imread(image_file)
transform = np.array([1.2, 0.3, -100, 0.5, 1.2, -150, 0, 0])

jax_affined = jax_affine_transform(image, transform)
np_affined = np_affine_transform(image, transform)
tf_affined = tf_affine_transform(image, transform)
torch_affined = torch_affine_transform(image, transform)

fig, ax_dict = plt.subplot_mosaic([["A", "B", "C", "D", "E"]], figsize=(12, 4))
ax_dict["A"].set_title("Original")
ax_dict["A"].imshow(image)
ax_dict["B"].set_title("JAX Affined")
ax_dict["B"].imshow(jax_affined)
ax_dict["C"].set_title("Numpy Affined")
ax_dict["C"].imshow(np_affined)
ax_dict["D"].set_title("TF Affined")
ax_dict["D"].imshow(tf_affined)
ax_dict["E"].set_title("Torch Affined")
ax_dict["E"].imshow(torch_affined)
fig.tight_layout(h_pad=0.1, w_pad=0.1)
plt.savefig("affine.png")

```

Visualization:
affine

File: keras_core/ops/image_test.py [hunk 5]
```diff
@@ -101,3 +113,86 @@ def test_resize(self, method, antialias, data_format):
...
         self.assertAllClose(ref_out, out, atol=0.3)
+
+    @parameterized.parameters(
+        [
+            ("bilinear", "constant", "channels_last"),
+            ("nearest", "constant", "channels_last"),
+            ("bilinear", "nearest", "channels_last"),
+            ("nearest", "nearest", "channels_last"),
+            # ("bilinear", "wrap", "channels_last"),  no wrap in torch
+            # ("nearest", "wrap", "channels_last"),  no wrap in torch
```
Comment (fchollet) [inline keras_core/ops/image_test.py:124]: You can use `if backend.backend() == "torch": pytest.skip("wrap fill_mode not support in torch")` so that at least it's tested when supported

File: keras_core/ops/image.py [hunk 6]
```diff
@@ -102,3 +102,133 @@ def resize(
...
+            gradients are not backpropagated into transformation parameters.
+            Note that `c0` and `c1` are only effective when using TensorFlow
+            backend and will be considered as `0` when using other backends.
+        method: Interpolation method. Available methods are `"nearest"`,
+            and `"bilinear"`. Defaults to `"bilinear"`.
+        fill_mode: Points outside the boundaries of the input are filled
+            according to the given mode. Available methods are `"constant"`,
+            `"nearest"` and `"reflect"`. Defaults to `"constant"`.
+        fill_value: Value used for points outside the boundaries of the input if
+            fill_mode=`"constant"`. Defaults to `0`.
```
Comment (fchollet) [inline keras_core/ops/image.py:176]: Move backticks: `fill_mode="constant"`

File: keras_core/ops/image.py [hunk 7]
```diff
@@ -102,3 +102,133 @@ def resize(
...
+            tensor of size N x 8. If one row of transform is
+            `[a0, a1, a2, b0, b1, b2, c0, c1]`, then it maps the output point
+            `(x, y)` to a transformed input point
+            `(x', y') = ((a0 x + a1 y + a2) / k, (b0 x + b1 y + b2) / k)`,
+            where `k = c0 x + c1 y + 1`. The transform is inverted compared to
+            the transform mapping input points to output points. Note that
+            gradients are not backpropagated into transformation parameters.
+            Note that `c0` and `c1` are only effective when using TensorFlow
+            backend and will be considered as `0` when using other backends.
+        method: Interpolation method. Available methods are `"nearest"`,
```
Comment (fchollet) [inline keras_core/ops/image.py:170]: I just noticed that we have a slight API inconsistency: this function as well as `resize` both use `method`, but our preprocessing layers use `interpolation`. Perhaps we should just switch everything (including `resize`) to `interpolation` since the preprocessing layers are much older and more widely used.

**Review by fchollet — COMMENTED** [review]
Excellent -- thanks for the polished docstring!

File: keras_core/ops/image_test.py (hunk 5 above)
Comment (james77777778) [reply keras_core/ops/image_test.py]: @fchollet 
When using `wrap`, the output in TF shows inconsistency compared to JAX and NumPy.
What make this situation awkward is that the test sometimes passes but at other times it fails.

As a solution, I have opted to skip the test to prevent CI failures.

File: keras_core/ops/image.py (hunk 6 above)
Comment (james77777778) [reply keras_core/ops/image.py]: Done

File: keras_core/ops/image.py (hunk 7 above)
Comment (james77777778) [reply keras_core/ops/image.py]: I have replaced `method` with `interpolation` in both `affine_transform` and `resize`.
Additionally, there have been some API changes in keras-core.

File: keras_core/ops/image_test.py (hunk 5 above)
Comment (james77777778) [reply keras_core/ops/image_test.py]: The mismatched elements is about 0.6%
```bash
Mismatched elements: 45 / 7500 (0.6%)  # one run for JAX
Mismatched elements: 48 / 7500 (0.64%)  # one run for NumPy
```

**Review by fchollet — APPROVED** [review]
Awesome -- thanks for the great contribution! 👍
